File: python/Rotate-List/Rotate-List.py
## Solution 1:
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if head==None or k<=0:
            return head
        temp = head
        n = 0
        while temp:
            n+=1
            temp = temp.next
        k = k%n
        temp = head
        for i in range(k):
            temp = self.rightMove(temp)
        return temp
    
    def rightMove(self,head):
        if head==None or head.next==None:
            return head
        prev = None
        cur = head
        temp = head
        while cur.next:
            prev = cur
            cur = cur.next
        prev.next = cur.next
        cur.next = temp
        return cur
## Solution 2:
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if head==None or k<=0:
            return head
        cur = head
        tail = None
        n = 0
        while cur:
            n+=1
            tail = cur
            cur = cur.next
        k = k%n
        tail.next = head
        cur = head
        for i in range(n-k-1):
            cur = cur.next
        head = cur.next
        cur.next = None
        return head

File: cpp/Rotate-List/Rotate-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head==NULL || k<=0)
            return head;
        ListNode* temp = head;
        int n = 0;//链表长度
        while(temp)
        {
            n++;
            temp = temp->next;
        }
        ListNode* dummyhead = new ListNode(-1);
        temp = head;
        k = k%n;
        for (int i=0;i<k;i++)
        {
            temp = rightMove(temp);
        }
        dummyhead->next = temp;
        return dummyhead->next;
    }
    ListNode* rightMove(ListNode* head)
    {
        if (head==NULL || head->next==NULL) 
            return head;
        ListNode* prev = NULL;
        ListNode* cur = head;
        ListNode* temp = head;
        while(cur->next!=NULL)
        {
            prev = cur;
            cur = cur->next;
        }
        prev->next = cur->next;
        cur->next = temp;
        return cur;
    }
};
